File: aio/coreys-party.py
partyin = open("partyin.txt", "r")
partyout = open("partyout.txt", "w")

n, m, a, b= map(int, partyin.readline().split())

nums = {}

for i in range(n):
    nums[i+1] = set()

for i in range(m):
    x, y = map(int, partyin.readline().split())
    nums[x].add(y)
    nums[y].add(x)

changed = True

while changed:
    changed = False
    toremove = set()
    for i in nums:
        if len(nums[i]) < a or n - len(nums[i]) - 1 < b:
            print("removing " + str(i))
            changed = True
            toremove.add(i)
    for i in toremove:
        if i in nums:
            nums.pop(i)
        for q in nums:
            if i in nums[q]:
                nums[q].remove(i)
        n -= 1

print(len(nums))

partyout.write(str(len(nums)))

File: aio/coreys-party.cpp
#include <map>
#include <set>
#include <vector>
#include <cstdio>
#include <iostream>

using namespace std;

int main() {
    freopen("partyin.txt", "r", stdin);
    freopen("partyout.txt", "w", stdout);

    int n, m, a, b, x, y;
    cin >> n >> m >> a >> b;
    
    map <int, set <int>> nums;
    set <int> emptyset;

    for (int i = 1; i <= n; i++) {
        nums[i] = emptyset;
    }

    for (int i = 0; i < m; i++) {
        cin >> x >> y;
        nums[x].insert(y);
        nums[y].insert(x);
    }

    bool changed = true;
    vector <int> to_remove;

    while (changed) {
        changed = false;
        for (auto it: nums) {
            if (it.second.size() < a || n - it.second.size() - 1 < b) {
                changed = true;
                to_remove.push_back(it.first);
            }
        }

        for (int i : to_remove) {
            if (nums.find(i) != nums.end()) {
                n--;
                for (int j : nums[i]) {
                    nums[j].erase(i);
                }
                nums.erase(i);
            }
        }
    }

    cout << n;

}
